gems/sorbet-runtime/test/types/final_method.rb

```diff
@@ -0,0 +1,319 @@
+    m1 = Module.new do
+      extend T::Sig
+      sig(:final) {void}
+      def foo; end
+    end
+    m2 = Module.new do
+      def foo; end
+    end
+    err = assert_raises(RuntimeError) do
+      Class.new do
+        include m2, m1
+      end
+    end
+    assert_includes(err.message, "was declared as final and cannot be overridden")
+  end
+
+  it "forbids overriding a final method by extending two modules" do
+    m1 = Module.new do
+      extend T::Sig
+      sig(:final) {void}
+      def foo; end
+    end
+    m2 = Module.new do
+      def foo; end
+    end
+    err = assert_raises(RuntimeError) do
+      Class.new do
+        extend m2, m1
+      end
+    end
+    assert_includes(err.message, "was declared as final and cannot be overridden")
+  end
+
+  it "allows calling final methods" do
+    m = Module.new do
+      extend T::Sig
+      sig(:final) {void}
+      def self.n0; end
+      sig(:final) {params(x: Integer).void}
+      def self.n1(x); end
+      sig(:final) {params(x: Integer, y: Integer).void}
+      def self.n2(x, y); end
+      sig(:final) {params(x: Integer, y: Integer, z: Integer).void}
+      def self.n3(x, y, z); end
+    end
+    m.n0
+    m.n1 1
+    m.n2 1, 2
+    m.n3 1, 2, 3
+  end
+
+  it "calls a user-defined included" do
+    m = Module.new do
+      @calls = 0
+      extend T::Sig
+      sig(:final) {returns(Integer)}
+      def self.calls
+        @calls
+      end
+      def self.included(x)
+        @calls += 1
+      end
+    end
+    Class.new do
+      include m
+    end
+    assert_equal(1, m.calls)
+    Class.new do
+      include m
+    end
+    assert_equal(2, m.calls)
+  end
+
+  it "calls a user-defined extended" do
+    m = Module.new do
+      @calls = 0
+      extend T::Sig
+      sig(:final) {returns(Integer)}
+      def self.calls
+        @calls
+      end
+      def self.extended(x)
+        @calls += 1
+      end
+    end
+    Class.new do
+      extend m
+    end
+    assert_equal(1, m.calls)
+    Class.new do
+      extend m
+    end
+    assert_equal(2, m.calls)
+  end
+
+  it "forbids overriding through many levels of include" do
+    m1 = Module.new do
+      extend T::Sig
+      sig(:final) {void}
+      def foo; end
+    end
+    m2 = Module.new do
+      include m1
+    end
+    m3 = Module.new do
+      include m2
+    end
+    err = assert_raises(RuntimeError) do
+      Class.new do
+        include m3
+        def foo; end
+      end
+    end
+    assert_includes(err.message, "was declared as final and cannot be overridden")
+  end
+
+  it "allows including modules again" do
+    m1 = Module.new do
+      extend T::Sig
+      sig(:final) {void}
+      def foo; end
+    end
+    m2 = Module.new do
+      include m1, m1
+    end
+  end
+
+  it "allows extending modules again" do
+    m1 = Module.new do
+      extend T::Sig
+      sig(:final) {void}
+      def foo; end
+    end
+    m2 = Module.new do
+      extend m1, m1
+    end
+  end
```

Comment:
For my understanding, what are these two "allows including/extending modules again" tests for?

Reply:
They're making sure that if we {include, extend} the same module twice, we don't incorrectly say we're overriding ourself.